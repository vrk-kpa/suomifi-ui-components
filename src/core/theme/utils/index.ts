export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
