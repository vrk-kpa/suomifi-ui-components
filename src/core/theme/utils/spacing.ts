import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

export type GlobalMarginProps = { globalMargins: GlobalMargins };

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export type GlobalMargins = {
  all?: MarginProps;
  alert?: MarginProps;
  actionMenu?: MarginProps;
  block?: MarginProps;
  button?: MarginProps;
  breadcrumb?: MarginProps;
  checkbox?: MarginProps;
  checkboxGroup?: MarginProps;
  chip?: MarginProps;
  dateInput?: MarginProps;
  dropdown?: MarginProps;
  expander?: MarginProps;
  expanderGroup?: MarginProps;
  externalLink?: MarginProps;
  fileInput?: MarginProps;
  heading?: MarginProps;
  hintText?: MarginProps;
  inlineAlert?: MarginProps;
  label?: MarginProps;
  languageMenu?: MarginProps;
  link?: MarginProps;
  linkList?: MarginProps;
  loadingSpinner?: MarginProps;
  multiSelect?: MarginProps;
  notification?: MarginProps;
  pagination?: MarginProps;
  paragraph?: MarginProps;
  radioButton?: MarginProps;
  radioButtonGroup?: MarginProps;
  routerLink?: MarginProps;
  searchInput?: MarginProps;
  serviceNavigation?: MarginProps;
  sideNavigation?: MarginProps;
  singleSelect?: MarginProps;
  staticChip?: MarginProps;
  statusText?: MarginProps;
  text?: MarginProps;
  textarea?: MarginProps;
  textInput?: MarginProps;
  timeInput?: MarginProps;
  toast?: MarginProps;
  toggleInput?: MarginProps;
  toggleButton?: MarginProps;
  tooltip?: MarginProps;
  wizardNavigation?: MarginProps;
};

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (
  spacing?: SpacingProps,
  important?: boolean,
): string => {
  if (!spacing) return '';

  const cssStyles = Object.entries(spacing)
    .map(([key, value]) =>
      getCSSSpacing(
        defaultSuomifiTheme,
        key as keyof SpacingProps,
        value,
        important,
      ),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
  important: boolean = false,
) => {
  const amount = spaceVal(theme)(value);
  const importantValue = important ? '!important' : '';
  switch (key) {
    case 'mx':
      return `margin-right: ${amount} ${importantValue}; margin-left: ${amount} ${importantValue};`;
    case 'my':
      return `margin-top: ${amount} ${importantValue}; margin-bottom: ${amount} ${importantValue};`;
    case 'px':
      return `padding-right: ${amount} ${importantValue}; padding-left: ${amount} ${importantValue};`;
    case 'py':
      return `padding-top: ${amount} ${importantValue}; padding-bottom: ${amount} ${importantValue};`;
    default:
      return `${cssSelector[key]}: ${amount} ${importantValue};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
