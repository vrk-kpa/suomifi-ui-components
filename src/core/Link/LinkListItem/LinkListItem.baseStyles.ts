import { SuomifiTheme } from '../../theme';
import { css } from 'styled-components';
import { font } from '../../theme/reset';

export const LinkListItemStyles = (theme: SuomifiTheme) => css`
  ${font(theme)('bodyText')}
  &.fi-link-list-item {
    list-style: none;
    padding: 0;
    & .fi-link-list-item_icon {
      margin-right: 2px;
      & .fi-icon {
        margin-left: -3px;
        font-size: 16px;
        transform: translateY(0.1em);
        & .fi-icon-base-fill {
          fill: ${theme.colors.accentBase};
        }
      }
    }
    & .fi-link {
      margin: 0;
    }
  }
`;
