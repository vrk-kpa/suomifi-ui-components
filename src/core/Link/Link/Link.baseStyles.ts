import { SuomifiTheme } from '../../theme';
import { css } from 'styled-components';
import { baseStyles } from '../BaseLink/BaseLink.baseStyles';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const LinkStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  ${baseStyles(theme)}
`;
