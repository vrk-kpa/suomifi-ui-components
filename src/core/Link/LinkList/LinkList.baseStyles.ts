import { SuomifiTheme } from 'core/theme';
import { css } from 'styled-components';
import { font } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const LinkListStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  &.fi-link-list {
    padding: 0;
    margin: 5px 0 0 0;
  }
  &.fi-link-list--small {
    & .fi-link-list-item_icon .fi-icon {
      transform: translateY(0.15em);
    }
    & .fi-link {
      ${font(theme)('bodyTextSmall')}
    }
  }
`;
