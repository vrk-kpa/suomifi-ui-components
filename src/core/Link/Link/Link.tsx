import React, { forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { IconChevronRight } from 'suomifi-icons';
import { LinkStyles } from '../Link/Link.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { HtmlA } from '../../../reset';
import {
  BaseLinkProps,
  baseClassName,
  linkClassNames,
} from '../BaseLink/BaseLink';
import { filterDuplicateKeys } from '../../../utils/common/common';

export interface LinkProps extends BaseLinkProps, MarginProps {
  /** Ref is forwarded to the anchor element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLAnchorElement>;
}

const StyledLink = styled(
  ({
    asProp,
    className,
    smallScreen,
    theme,
    globalMargins,
    variant = 'default',
    children,
    underline = 'hover',
    ...rest
  }: LinkProps & SuomifiThemeProp & GlobalMarginProps) => {
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlA
        {...passProps}
        className={classnames(baseClassName, className, {
          [linkClassNames.linkUnderline]: underline === 'initial',
          [linkClassNames.accent]: variant === 'accent',
          [linkClassNames.small]: smallScreen,
        })}
        as={asProp}
        style={{ ...passProps?.style }}
      >
        {variant === 'accent' && (
          <IconChevronRight
            color={theme.colors.accentBase}
            className={linkClassNames.accentIcon}
          />
        )}
        {children}
      </HtmlA>
    );
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.button,
      marginProps,
    );
    return LinkStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Link = forwardRef(
  (props: LinkProps, ref: React.Ref<HTMLAnchorElement>) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledLink
              theme={suomifiTheme}
              globalMargins={margins}
              forwardedRef={ref}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

Link.displayName = 'Link';
export { Link };
