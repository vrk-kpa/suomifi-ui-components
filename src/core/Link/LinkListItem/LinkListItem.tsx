import React, { forwardRef, ReactElement } from 'react';
import { styled } from 'styled-components';
import classnames from 'classnames';
import { IconChevronRight } from 'suomifi-icons';
import { LinkListItemStyles } from './LinkListItem.baseStyles';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { HtmlLi, HtmlLiProps, HtmlSpan } from '../../../reset';
import { LinkProps } from '../Link/Link';
import { ExternalLinkProps } from 'core/Link/ExternalLink/ExternalLink';
import { RouterLinkProps } from 'core/Link/RouterLink/RouterLink';

const baseClassName = 'fi-link-list-item';

const listLinkClassNames = {
  icon: `${baseClassName}_icon`,
};

export interface LinkListItemProps extends Omit<HtmlLiProps, 'className'> {
  /** Ref is forwarded to the anchor element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLLIElement>;
  /** Custom icon from suomifi-icons to use for the list item */
  icon?: ReactElement;
  /** The actual link element of the list item */
  children: ReactElement<
    LinkProps | ExternalLinkProps | RouterLinkProps<React.ElementType>
  >;
  /** CSS class for custom styles */
  className?: string;
}

const StyledLinkListItem = styled(
  ({
    children,
    theme,
    className,
    icon,
    ...passProps
  }: LinkListItemProps & SuomifiThemeProp) => (
    <HtmlLi {...passProps} className={classnames(baseClassName, className)}>
      <HtmlSpan className={listLinkClassNames.icon}>
        {!!icon ? icon : <IconChevronRight />}
      </HtmlSpan>
      {children}
    </HtmlLi>
  ),
)`
  ${({ theme }) => LinkListItemStyles(theme)}
`;

/**
 * <i class="semantics" />
 * Used as a wrapper for links in LinkList to add correct styling.
 */
const LinkListItem = forwardRef(
  (props: LinkListItemProps, ref: React.Ref<HTMLElement>) => (
    <SuomifiThemeConsumer>
      {({ suomifiTheme }) => (
        <StyledLinkListItem
          theme={suomifiTheme}
          forwardedRef={ref}
          {...props}
        />
      )}
    </SuomifiThemeConsumer>
  ),
);

LinkListItem.displayName = 'LinkListItem';
export { LinkListItem };
