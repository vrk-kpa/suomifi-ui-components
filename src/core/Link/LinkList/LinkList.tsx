import React, { forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { LinkListStyles } from './LinkList.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { HtmlUlProps, HtmlUlWithRef } from '../../../reset';
import { getConditionalAriaProp } from '../../../utils/aria';
import { filterDuplicateKeys } from '../../../utils/common/common';

const LinkListClassName = 'fi-link-list';
const SmallScreenClassName = 'fi-link-list--small';

export interface LinkListProps extends HtmlUlProps, MarginProps {
  /** Ref is forwarded to the list element. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLUListElement>;
  /** Id of the heading or label of the list */
  ariaLabelledBy: string;
  /** Sets smaller font size for the list elements */
  smallScreen?: boolean;
}

const StyledLinkList = styled(
  ({
    className,
    theme,
    children,
    smallScreen,
    ariaLabelledBy,
    globalMargins,
    forwardedRef,
    ...rest
  }: LinkListProps & SuomifiThemeProp & GlobalMarginProps) => {
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlUlWithRef
        ref={forwardedRef}
        {...passProps}
        className={classnames(className, LinkListClassName, {
          [SmallScreenClassName]: smallScreen,
        })}
        {...getConditionalAriaProp('aria-labelledby', [ariaLabelledBy])}
        style={{ ...passProps?.style }}
      >
        {children}
      </HtmlUlWithRef>
    );
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.linkList,
      marginProps,
    );
    return LinkListStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const LinkList = forwardRef(
  (props: LinkListProps, ref: React.Ref<HTMLUListElement>) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledLinkList
              theme={suomifiTheme}
              globalMargins={margins}
              forwardedRef={ref}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

LinkList.displayName = 'LinkList';
export { LinkList };
