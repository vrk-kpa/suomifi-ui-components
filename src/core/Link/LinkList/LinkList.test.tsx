import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';

import { LinkListItem } from '../LinkListItem/LinkListItem';
import { LinkList } from './LinkList';
import { Link } from '../Link/Link';

const TestLinkList = (
  <>
    <h1 id="a">Heading</h1>
    <LinkList ariaLabelledBy="a">
      <LinkListItem data-testid="test-link">
        <Link href="/">Link</Link>
      </LinkListItem>
      <LinkListItem data-testid="test-link1">
        <Link href="/">AA Link</Link>
      </LinkListItem>
    </LinkList>
  </>
);

describe('Simple link list with one item', () => {
  it('Should be unordered', () => {
    const { getAllByRole } = render(TestLinkList);
    const firstLink = getAllByRole('link')[0];
    expect(firstLink.textContent).toBe('Link');
  });

  it('should match snapshot', () => {
    const { baseElement } = render(TestLinkList);
    expect(baseElement).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', axeTest(TestLinkList));
});

describe('Margin prop', () => {
  it('should have margin style from margin prop', () => {
    const { baseElement } = render(<LinkList ariaLabelledBy="" margin="xs" />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should have margin prop overwritten from style prop', () => {
    const { container } = render(
      <LinkList ariaLabelledBy="" margin="xs" style={{ margin: 2 }} />,
    );
    expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
  });
});
