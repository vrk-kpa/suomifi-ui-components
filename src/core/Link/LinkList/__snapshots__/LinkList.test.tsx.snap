// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Margin prop should have margin style from margin prop 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c1 {
  margin: 10px !important;
}

.c1.fi-link-list {
  padding: 0;
  margin: 5px 0 0 0;
}

.c1.fi-link-list--small .fi-link-list-item_icon .fi-icon {
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c1.fi-link-list--small .fi-link {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

<body>
  <div>
    <ul
      class="c0 c1 fi-link-list"
    />
  </div>
</body>
`;

exports[`Simple link list with one item should match snapshot 1`] = `
.c5 {
  vertical-align: baseline;
}

.c5.fi-icon {
  display: inline-block;
}

.c5 .fi-icon-base-fill {
  fill: currentColor;
}

.c5 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c5.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c5.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c6:hover,
.c6:active,
.c6:focus,
.c6:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c3 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c3.fi-link-list-item {
  list-style: none;
  padding: 0;
}

.c3.fi-link-list-item .fi-link-list-item_icon {
  margin-right: 2px;
}

.c3.fi-link-list-item .fi-link-list-item_icon .fi-icon {
  margin-left: -3px;
  font-size: 16px;
  -webkit-transform: translateY(0.1em);
  -ms-transform: translateY(0.1em);
  transform: translateY(0.1em);
}

.c3.fi-link-list-item .fi-link-list-item_icon .fi-icon .fi-icon-base-fill {
  fill: hsl(23,82%,50%);
}

.c3.fi-link-list-item .fi-link {
  margin: 0;
}

.c1.fi-link-list {
  padding: 0;
  margin: 5px 0 0 0;
}

.c1.fi-link-list--small .fi-link-list-item_icon .fi-icon {
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c1.fi-link-list--small .fi-link {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c7 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c7.fi-link {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:hover,
.c7.fi-link:active,
.c7.fi-link:focus,
.c7.fi-link:focus-within {
  color: hsl(212,63%,45%);
}

.c7.fi-link:focus,
.c7.fi-link:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:focus {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
  outline: 3px solid transparent;
}

.c7.fi-link:hover,
.c7.fi-link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link:visited {
  color: hsl(284,36%,45%);
}

.c7.fi-link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:focus,
.c7.fi-link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:hover,
.c7.fi-link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link.fi-link--accent .fi-link--accent_icon {
  font-size: 16px;
  margin-left: -3px;
  margin-right: 2px;
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c7.fi-link.fi-link--small {
  font-size: 16px;
}

<body>
  <div>
    <h1
      id="a"
    >
      Heading
    </h1>
    <ul
      aria-labelledby="a"
      class="c0 c1 fi-link-list"
    >
      <li
        class="c2 fi-link-list-item c3"
        data-testid="test-link"
      >
        <span
          class="c4 fi-link-list-item_icon"
        >
          <svg
            aria-hidden="true"
            class="fi-icon c5"
            focusable="false"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fi-icon-base-fill"
              d="M15.61 11.01c.26.273.39.632.39.99s-.13.717-.39.99l-5.334 5.6a1.287 1.287 0 0 1-1.885 0 1.448 1.448 0 0 1 0-1.98l4.39-4.61-4.39-4.61a1.448 1.448 0 0 1 0-1.98 1.287 1.287 0 0 1 1.885 0l5.333 5.6Z"
              fill="#222"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <a
          class="c6 fi-link c7"
          href="/"
        >
          Link
        </a>
      </li>
      <li
        class="c2 fi-link-list-item c3"
        data-testid="test-link1"
      >
        <span
          class="c4 fi-link-list-item_icon"
        >
          <svg
            aria-hidden="true"
            class="fi-icon c5"
            focusable="false"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="fi-icon-base-fill"
              d="M15.61 11.01c.26.273.39.632.39.99s-.13.717-.39.99l-5.334 5.6a1.287 1.287 0 0 1-1.885 0 1.448 1.448 0 0 1 0-1.98l4.39-4.61-4.39-4.61a1.448 1.448 0 0 1 0-1.98 1.287 1.287 0 0 1 1.885 0l5.333 5.6Z"
              fill="#222"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <a
          class="c6 fi-link c7"
          href="/"
        >
          AA Link
        </a>
      </li>
    </ul>
  </div>
</body>
`;
